### PR TITLE
RSX: Bug fixes and improvements

### PR DESCRIPTION
**Pull request by kd-11**
- Fixes for backbuffer blits and complete removal of the flip hack when using openGL
- Implementation of blit engine in hardware for vulkan backend
- Reorganization of frame storage and heap management for vulkan [WIP, slight performance degradation]
- RSX thread recovery in the event of an unknown command [WIP, still dies/loops infinitely if unmapped addresses are provided]
- Unified texture cache for vulkan and openGL. This means that any bugs fixed for one backend should carry over making development easier
- Removal of redundant classes in openGL and general cleanup of the texture framework for that backend. There were 2 separate texture classes for example with very different approaches which was awful
- Implements draw call batching for the slow games that would throw thousands of small draws at the hardware. Significantly speeds some some games. There is an AMD driver bug that will cause missing graphics but I have reported this upstream.
- Adds support for split draw ranges which would otherwise assert in RSXThread
- Fixes for some shader decompiler bugs (DBZ Raging Blast games)
- Improve mesa compatibility for openGL

This is alot of new code so take your time testing. Try not to report duplicates to make my work a little easier.

### CODE REVIEW

**Comment by alex88510** [comment]
With this PR, Trails of Cold Steel shows black screen with GPU texture scaling ON. 
This game needs GPU texture scaling or Write Color Buffer to shows graphics properly. 

capture

[RPCS3.log1.gz](https://github.com/RPCS3/rpcs3/files/1309119/RPCS3.log1.gz)

Vulkan with Write Color Buffer ON is still showing black screen.

**Comment by Sylfayn** [comment]
With this, Persona 4 Arena shows a black screen on both Vulkan and OpenGL and seems to freeze completely after the opening video in OpenGL.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1309133/RPCS3.log.gz)

**Comment by Illynir** [comment]
Nier doesn't works on this PR, black screen with 0 FPS and nothing. On OGL and Vulkan, with and without GPU scaling. With and without shader cache.

[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1309138/RPCS3.zip)

Emulator is also very unstable, crash often, always in background after closing, etc...

**Comment by Illynir** [comment]
ON KH 1 with GPU texture scaling ON with Vulkan:

image

Without it's good however but i'm saying it just in case.

**Comment by MarioSonic2987** [comment]
image
image
image

Fixes blackscreen regression in Worms 2: Armageddon, Worms Revolution and Worms Crazy Golf.

image
But with GPU scaling OFF, there are glitched textures (2D).

**Comment by MarioSonic2987** [comment]
`E {rsx::thread} RSX: No valid program bound to pipeline. Skipping draw`

Spams this in Worms Ultimate Mayhem with Vulkan.
image

**Comment by Renusek** [comment]
NPEB00388 shows more now, but still glitched.
obraz

and NPUB30231 is showing a black screen instead of a cutscene before the gameplay now

**Comment by Lardian** [comment]
Demon's Souls [BLUS30443] gets error right after it loading 
F {PPU[0x1000005] Thread (HighGraphics) [0x00a662f0]} class std::runtime_error thrown: Verification failed: 
(in file c:\rpcs3\rpcs3\emu\rsx\vk\../Common/texture_cache.h:59)

And error, after starting a new game.
F {rsx::thread} class std::runtime_error thrown: Verification failed: 
(in file c:\rpcs3\rpcs3\emu\rsx\vk\../Common/texture_cache.h:59)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1309175/RPCS3.log.gz)

**Comment by HenningJW** [comment]
I'm experiencing a pretty big performance regression in Catherine (BLES01459) with exactly the same settings on an 2500k@4GHz HD7850 Vulkan (1 SPU Thread, SPU Loop Detection ON, 0 SPU Delay Penalty, Strict Rendering Mode).

Screenshot taken on the second night in the bar, the first one shows 30fps and is smooth, but a slight adjustment to the camera (right analog stick) drops frames significantly like in the second screenshot. This happen in a lot of places and is repeatable. On master everything is at 30fps no matter where. I am not sure if the amount of object cause the drop or specific types of objects.

image

image

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1309287/RPCS3.log.gz)

If you want further testing or you want a savegame I am happy to assist.

**Comment by AniLeo** [comment]
Lord of the Rings Conquest regressed, now dies right before character select on both backends with

F {rsx::thread} class std::runtime_error thrown: Verification failed (e=0x1e7): 
(in file C:\rpcs3\Utilities\VirtualMemory.cpp:83)

**Comment by raveskirza** [comment]
In the fbo-fixes branch before the sep17 changes were committed, Soul Calibur II HD online was completely playable as it was running smoothly in OpenGl and vulkan without Color buffers and strict rendering mode
This was my appveyor build with fbo-fixes which played the game smoothly before sep 17 commits
https://ci.appveyor.com/project/raveskirza/rpcs3/build/0.0.3-18/artifacts

But after the sep17 commits the Open Gl and vulkan show black screen with/without color buffers and Strict rendering on/off

Unfortunately I deleted my rpcs3 repository and forked it again today so I do not have the rpcs3 snapshot used for my above build.
Attached log from my appveyor build with Fbo-fixes from sep 14 and the lastest fbo-fixes of sep17 from the main rpcs3 appveyor build

[RPCS3_sep14_ fbo-fixes.txt](https://github.com/RPCS3/rpcs3/files/1309459/RPCS3_sep14_.fbo-fixes.txt)
[RPCS3_sep17_ fbo_fixes.txt](https://github.com/RPCS3/rpcs3/files/1309460/RPCS3_sep17_.fbo_fixes.txt)

**Comment by CesarAxe** [comment]
I tried SAW II Flesh and Blood [BLES01050] and got this error during the game: (Vulkan)

F {rsx::thread} class std::runtime_error thrown: Working buffer not big enough, buffer_length=67108864 allocated=63470080 requested=9296 guard=65536 largest_pool=10240
(in file c:\rpcs3\rpcs3\emu\rsx\d3d12\../Common/ring_buffer_helper.h:74)

In the current master branch I do not get this problem.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1309530/RPCS3.log.gz)

File: rpcs3/Emu/RSX/Common/surface_store.h [hunk 1]
```diff
@@ -437,5 +478,219 @@ namespace rsx
...
+					}
+				}
+			}
+		}
+
+		/**
+		 * Clipping and fitting lookup funcrions
+		 * surface_overlaps - returns true if surface overlaps a given surface address and returns the relative x and y position of the surface address within the surface
+		 * address_is_bound - returns true if the surface at a given address is actively bound
+		 * get_surface_subresource_if_available - returns a sectiion descriptor that allows to crop surfaces stored in memory
```
Comment (hcorion) [inline rpcs3/Emu/RSX/Common/surface_store.h]: A sectiion of this comment seems slightly off. 😆

**Comment by Zardas81** [comment]
Sword Art Online: Lost Song (BLAS50790): Has the same "4 mini windows" as some other games, maybe the extra log helps. Looks pretty much the same regardless of settings.
clipboard-1
[RPCS3.log1.gz](https://github.com/RPCS3/rpcs3/files/1309892/RPCS3.log1.gz)

**Comment by spring568** [comment]
Nice! Fighting climax works again in opengl! Though with GPU scaling on it looks sort of like that Worms 2: Armageddon screen shot with the distorted textures. Turning it off restores the games half missing characters and the menu textures are fine

**Comment by GeniusMage** [comment]
Tales of Vesperia has new crashes on Vulkan.
F {MFC Thread} class std::runtime_error thrown: Verification failed (e=0x57): 
(in file c:\rpcs3\rpcs3\emu\rsx\vk\../Common/texture_cache.h:59) (with GPU exture scaling)
F {SPU[0x2000001] Thread (SpursTLGraphicsCellSpursKernel1)} class std::runtime_error thrown: Verification failed (e=0x57): 
(in file c:\rpcs3\rpcs3\emu\rsx\vk\../Common/texture_cache.h:59) (without GPU texture scaling)

Seems to thave the same cause as Lardian's crash in Demon's Souls.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1310590/RPCS3.log.gz)

**Comment by bjakja** [comment]
On Fuel i noticed on last build some black and blue glitches. Game freezes after 2 minutes with 
E {rsx::thread} HLE: Getting the real address failed. (addr=0xffff368c)
E {rsx::thread} RSX: Invalid FIFO queue get/put registers found, get=0xFFFF368C, put=0x1F7FF4
This build also give me fps impact about 40%.

**Comment by Illynir** [comment]
New error on nier at launch with last commit: 

F {rsx::thread} class std::runtime_error thrown: Verification failed: 
(in file c:\rpcs3\rpcs3\emu\rsx\vk\../Common/texture_cache.h:59)

Without GPU scaling.

**Comment by Sylfayn** [comment]
Persona 4 Arena now shows graphics on Vulkan with a small black bar at the top and no longer freezes after the opening, however the entire emulator now ends up freezing once you hit the loading screen for the main menu. OpenGL now shows graphics again, but it still freezes after the opening.
p4avulkan

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1311678/RPCS3.log.gz)

**Comment by Zardas81** [comment]
That fixed the P4A Vulkan crashes. The game still crashes on OpenGL though. One glitch haven't seen mentioned before: a purple triangle appears instead for exactly one frame in each character's animation.
2017-09-18 22_35_39-fps_ 60 61 _ vulkan _ 0 0 3-5880-head _ persona 4 arena _ blus30985
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1312140/RPCS3.log.gz)

**Comment by Zardas81** [comment]
SAO: view is mostly fixed now. There are two black bars on left and right, and colors are weird.
2017-09-18 23_00_46-fps_ 19 81 _ vulkan _ 0 0 3-5880-head _ sword art online -lost song- _ blas5079
It's supposed to look like this:
2017-09-18 23_07_02-sword art online_ lost song _ episode 1 _ press start return to alo - youtube
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1312248/RPCS3.log.gz)

**Comment by SakataGintokiYT** [comment]
@Zardas81 
Stretch To Display  Area -> ON

**Comment by greentop** [comment]
Ubuntu 16.04 / Nvidia 384.69 / LLVM 4 / (PR3458 as of 188381c)

RPCS3 GPU Additional Settings:
Enabled - Invalidate Cache Every Frame & Use GPU Texture Scaling

Game: ToCS BLUS31572

PR3458 Vulkan: The right half of the game renders black. RPCS3 drags the OS to a crawl, requiring an forced end process to return Ubuntu to normal operation.
PR3458 OpenGL: The right half of the game renders black. 

Game: ToCS2 BLUS31597

PR3458 Vulkan: The right half of the game renders black. RPCS3 hangs shortly after game start beyond the save load. Very low fps observed.
PR3458 OpenGL: The right half of the game renders black. The snow fall effects are now working. The RPCS3 process sometimes does not end after closing the RPCS3 window, which requires the user to end the process manually.

[PR3458 RPCS3 logs and Screenshots.zip](https://github.com/RPCS3/rpcs3/files/1312358/PR3458.RPCS3.logs.and.Screenshots.zip)

tocs_vulkan_pr3458

**Comment by Sylfayn** [comment]
@Zardas81 Odd, my rip of Persona 4 Arena still crashes on Vulkan on the same loading screen, I can only guess it's my rip or something to do with my processor or graphics card. Strange.

**Comment by Asinin3** [comment]
**Gattling Gears**
`F {rsx::thread} class std::runtime_error thrown: Verification failed: 
(in file c:\rpcs3\rpcs3\emu\rsx\vk\../Common/texture_cache.h:131)`
Getting this error when trying to launch the game with this PR, doesn't happen in master.

**Comment by raveskirza** [comment]
@Asinin3 Disable Write color buffers on

**Comment by Asinin3** [comment]
@raveskirza Yes, I know but it does not happen with master I already updated my comment do indicate that.

**Comment by raveskirza** [comment]
@Asinin3 if you use my build posted above. probably you may not need write color buffers on. I believe the fbo-fixes should eliminate the need to have "Write color buffers". But right now if you uncheck write color buffers it just shows black screen. Some commit on sep17 in the fbo-fixes branch broke the behavior of disabling write color buffers on

**Comment by Kojikun1** [comment]
Nier Replicant  - Vulkan 
F {rsx::thread} class std::runtime_error thrown: Verification failed (e=0x7e): 
(in file C:\rpcs3\rpcs3\Emu\RSX\VK\VKGSRender.cpp:2353)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1312618/RPCS3.log.gz)

**Comment by Kojikun1** [comment]
Nier Replicant - Vulkan 
F {rsx::thread} class std::runtime_error thrown: Verification failed: 
(in file c:\rpcs3\rpcs3\emu\rsx\vk\../Common/texture_cache.h:131)

**Comment by twdarkeh** [comment]
http://puu.sh/xDc5d.jpg

The half black screen happens in Akiba's Trip as well, with or without invalidate texture cache.

**Comment by SakataGintokiYT** [comment]
**[Vulkan]**
Kung Fu Panda 2 closes emulator after startup
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1313688/RPCS3.zip)

**Comment by alex88510** [comment]
On commit [0b5bf51f](https://github.com/RPCS3/rpcs3/commit/0b5bf51fa75c9c367417a585c4311f5dee06a1d6), Trails of Cold Steel shows around 1/3 of the screen with both openGL and vulkan, GPU texture scaling on and off. 

capture2

[RPCS3_openglNoScaling.gz](https://github.com/RPCS3/rpcs3/files/1314063/RPCS3_openglNoScaling.gz)

Using vulkan, the emulator crashed very often(almost every time before reaching the main menu). 
But once it did manage to get into the game, I see better fps than with opengl.

**Comment by Sandvichero** [comment]
Nier Replicant- Vulcan
Now goes ingame with numerous new bugs

It spams These in the log:
E {PPU[0x1000000] Thread (main_thread) [0x00f7d450]} RSX: Invalid bpp 64
E {PPU[0x1000000] Thread (main_thread) [0x00f7d450]} RSX: Invalid bpp 16
E {PPU[0x1000000] Thread (main_thread) [0x00f7d450]} RSX: Invalid bpp 32
E {PPU[0x1000000] Thread (main_thread) [0x00f7d450]} RSX: Invalid bpp 64
E {PPU[0x1000000] Thread (main_thread) [0x00f7d450]} RSX: Invalid bpp 16
E {PPU[0x1000000] Thread (main_thread) [0x00f7d450]} RSX: Invalid bpp 32
E {PPU[0x1000000] Thread (main_thread) [0x00f7d450]} RSX: Invalid bpp 64
E {rsx::thread} RSX: Invalid bpp 16
E {rsx::thread} RSX: Invalid bpp 32

Also it now crashes regularly with
E {PPU[0x1000016] Thread (_cellsurMixerMain) [0x008a2a6c]} Stat: '_sys_lwmutex_lock' failed with 0x80010005 : CELL_ESRCH [x113388]

The game is much darker now as well
Before:
capteure

After:
capture

Log: 
[RPCS3.log1.gz](https://github.com/RPCS3/rpcs3/files/1314655/RPCS3.log1.gz)

**Comment by kd-11** [comment]
Wcb+vulkan is broken in select cases. Its a minor bug and will be fixed soon.
Whats important right now is stopping the crashes (purple text crashes or video driver crashes or god forbid BSOD)

**Comment by OverlordCW** [comment]
RDR Goty [BLES01294]
rdr

And:
`E {rsx::thread} RSX: Invalid RSX method 0x4 (arg=0xc1560000)
(in file C:\rpcs3\rpcs3\Emu\RSX\rsx_methods.cpp:45)
E {rsx::thread} RSX: Invalid RSX method 0x80 (arg=0x1760080)
(in file C:\rpcs3\rpcs3\Emu\RSX\rsx_methods.cpp:45)`
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1314756/RPCS3.log.gz)

**Comment by SNK-Afridi** [comment]
Mortal Kombat Bles01132 show screen zoomed most of the time using latest build  [049f1954](https://github.com/RPCS3/rpcs3/commit/049f19545915ab6718912c0bb5065342f8efc2b3) and crashes with
F {rsx::thread} class std::runtime_error thrown: Unknown index array type 116
(in file C:\rpcs3\rpcs3\Emu\RSX\gcm_enums.cpp:26)

desktop screenshot 2017 09 19 - 20 37 19 09
desktop screenshot 2017 09 19 - 20 37 04 59
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1314852/RPCS3.log.gz)

**Comment by Luffykun007** [comment]
OpenGL: 

capturar1

Vulkan:

capturar

**Comment by toccata10** [comment]
Nice improvement on Raiden IV, on linux (r9 fury, mesa 17.2.0). No longer crashes, much better graphics (still some glitches, but nice progress).
[RPCS3-RaidenIV.zip](https://github.com/RPCS3/rpcs3/files/1316923/RPCS3-RaidenIV.zip)
fps 60 00 opengl 0 0 3-5917 raiden iv overkill npub31371 _006

fps 60 01 opengl 0 0 3-5917 raiden iv overkill npub31371 _005
fps 60 48 opengl 0 0 3-5917 raiden iv overkill npub31371 _004

**Comment by OverlordCW** [comment]
RDR Goty [BLES01294]
`E {rsx::thread} RSX: NV3089_IMAGE_IN_SIZE: Invalid blit dimensions passed`
`F {rsx::thread} class std::runtime_error thrown: Unexpected vkFormat 0x85`
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/1317981/RPCS3.log.gz)

**Comment by twdarkeh** [comment]
F {rsx::thread} class std::runtime_error thrown: Assertion failed! Result is FFFFFFFCh
(in file c:\rpcs3\rpcs3\emu\rsx\vk\VKHelpers.h:1070)

In Akiba's Trip.

http://puu.sh/xEsKB.zip Log.

**Comment by kd-11** [comment]
If its consistently triggered, enable debug output (if you have the vulkan sdk) and enable strict mode rendering then run. The log will then have validation output which will be useful in debugging the driver crash.

**Comment by aakh1361** [comment]
RPCS3 v0.0.3-5885-0a3d0da7 Alpha | HEAD

OpenGL
image
image


[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1318194/RPCS3.zip)

**Comment by aakh1361** [comment]
RPCS3 v0.0.3-5885-0a3d0da7 Alpha | HEAD
God of War® Collection

[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1318270/RPCS3.zip)

**Comment by kd-11** [comment]
Is that GO1, GOW2 or the launcher? Coz GOW1 and 2 work fine

**Comment by aakh1361** [comment]
In launcher
when select File "GAMESEL.self"

**Comment by kd-11** [comment]
Are you running the latest build? Just checked again and its fine - even with strict rendering mode

**Comment by toccata10** [comment]
Virtual fighter 5[NPEB00913]
Linux, r9 fury, mesa. Ogl: some artifacts all over the image (rapidly changing). Tried with strict rendering mode:same pb. 
Vulkan: Freezes before starting a fight.
fps 1 50 opengl 0 0 3-5924 virtua fighter 5 final showdown npeb00913 _017
fps 10 35 opengl 0 0 3-5924 virtua fighter 5 final showdown npeb00913 _016
fps 24 66 opengl 0 0 3-5924 virtua fighter 5 final showdown npeb00913 _018
[RPCS3.log-VF5-OGL-glitches.gz](https://github.com/RPCS3/rpcs3/files/1318445/RPCS3.log-VF5-OGL-glitches.gz)
[RPCS3.log-VF5-Vulkan-Freeze.zip](https://github.com/RPCS3/rpcs3/files/1318446/RPCS3.log-VF5-Vulkan-Freeze.zip)

**Comment by aakh1361** [comment]
Opps When Change Resoultion Game Of 720_480 To 1280_720
Now Everything "God of war Collection" Fix :-)
image

But Resident Evil Code veronica X don't fix
image

[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/1318439/RPCS3.zip)

**Comment by bjakja** [comment]
Sengoku Basara 4 Sumeragi do not work without white color buffer anymore, and it show 2/3 of screen exactly like Resident Evil Code veronica X.
clipboard01
clipboard02
On second screen there is also blue bar on the right side.

**Comment by Zardas81** [comment]
@Sylfayn Now that I played it further, some videos do crash for me too. But not all of them. It seems to depend on length, short scenes are usually fine, but longer ones crash.

**Comment by kd-11** [comment]
Retest games with clipping bugs (cropped output)

**Comment by Illynir** [comment]
image

In some areas, thereis the issue on Nier. Its still better that on Master or it's black screen.

Open and close menu fix the problem for some seconds only, the problem come back immediatly.

image

image

With and without GPU scaling, no difference. GPU make the game freeze on loading screen sometimes.

**Comment by Illynir** [comment]
image

Nier Replicant has the problem almost everywhere also, more that US version of Nier.

**Comment by bjakja** [comment]
Sengoku Basara 4 Sumeragi is fixed with latest build but Legend of Heroes Trails of Cold Steel I is still show a 1/3 of screen.

**Comment by aakh1361** [comment]
fixed "resident evil code veronica x"
image

**Comment by Zardas81** [comment]
Persona 4 Arena: Output regressed to this on both renderers, and also instantly kills the program when entering the main menu.
2017-09-20 22_01_56-fps_ 60 85 _ vulkan _ 0 0 3-5890-head _ persona 4 arena _ blus30985
[RPCS3.ZIP](https://github.com/RPCS3/rpcs3/files/1319064/RPCS3.ZIP)

**Comment by LonelyDragon123** [comment]
qavhh9b
m1kpyk7
kaofygf

**Comment by greentop** [comment]
Ubuntu 16.04 / Nvidia 384.69 / LLVM 4 / (PR3458 as of 0e6b5c8)

RPCS3 GPU Additional Settings:
Enabled - Invalidate Cache Every Frame & Use GPU Texture Scaling

Game: ToCS BLUS31572

PR3458 Vulkan: The game window hangs upon loading the intro video. RPCS3 process required manual end process after closing RPCS3 main window. (This is repeatable.)
PR3458 OpenGL: The game can go past the intro video, but shortly after at the title screen that OS drags to a crawl. An end process of RPCS3 was done to return Ubuntu back to normal operation. The right half of the game renders black. Unable to acquire screenshot.

Game: ToCS2 BLUS31597

PR3458 Vulkan: The game window hangs upon loading the intro video. RPCS3 process required manual end process after closing RPCS3 main window. (This is repeatable.)
PR3458 OpenGL: The game can go past the intro video, but shortly after at the title screen that OS drags to a crawl. An end process of RPCS3 was done to return Ubuntu back to normal operation. The right half of the game renders black. Unable to acquire screenshot.

[PR3458_0e6b5c8_logs.zip](https://github.com/RPCS3/rpcs3/files/1319259/PR3458_0e6b5c8_logs.zip)

**Comment by kd-11** [comment]
The bug with missing blit regions, etc (partly black) is fixed as of the latest commit https://github.com/RPCS3/rpcs3/pull/3458/commits/6915daa07496ea027d54418ff5ae68e296120e1c

**Comment by Luffykun007** [comment]
KD-11 I'm having this error now with Dengeki Bunko Fighting Climax, under-night in birth exe: late and Earthworm Jim HD, all with the same error, they will no longer play for me in both vulkan and opengl before it was good and I'm having black screen still with soul calibur 2 HD online.

capturar

capturar1

**Comment by JC-96** [comment]
@Luffykun007 I think that error is caused by the path being too long, try and change the location of your RPCS3 folder to a folder under C:\ or whatever hard drive you use.

**Comment by Luffykun007** [comment]
@JC-96 I will try this but before it was good.

**Comment by greentop** [comment]
**Ubuntu 16.04 / Nvidia 384.69 / LLVM 4 / (PR3458 as of cc77c58)**

**RPCS3 GPU Additional Settings:**
Enabled - Invalidate Cache Every Frame & Use GPU Texture Scaling

**Game: ToCS BLUS31572**

**PR3458 Vulkan:** The right side of the game window that was rendering black is fixed. Shadows are working. The game video hangs when accessing the Camp Menu and Quick Travel menu. An end process for RPCS3 was required after closing the RPCS3 main window. Game performance improved to a near consistent 30FPS in areas where there are no NPCs. (OpenGL on Master is 14 to 26 FPS.) Random game video freezing observed under idle. Note: Strict Rendering Mode will hang the game at the intro video. 
**PR3458 OpenGL:** The right side of the game window that was rendering black is fixed. Shadows are working. FPS appears to be the same as on Master. The Camp Menu and Quick Travel menu is working. RPCS3 closes out normally upon exiting the main RPCS3 window.

**Game: ToCS2 BLUS31597**

**PR3458 Vulkan:** The right side of the game window that was rendering black is fixed. Game freezing and very low FPS observed into the single digits. Purple lined at:
```
F {rsx::thread} St13runtime_error thrown: Assertion failed! Result is FFFFFFFEh
(in file /home/green/rpcs3/rpcs3/Emu/RSX/VK/VKHelpers.h:277)
```
**PR3458 OpenGL:** The right side of the game window that was rendering black is fixed. Shadows are working. FPS appears to be the same as on Master. The Camp Menu and Quick Travel menu is working. RPCS3 closes out normally upon exiting the main RPCS3 window.

Note: RPCS3 is allowing for multiple instances after a game freeze or crash. A check before RPCS3 startup should be done to ensure that only one instance of RPCS3 is running.
[RPCS3_PR3458_cc77c58_logs_and_screenshots.zip](https://github.com/RPCS3/rpcs3/files/1319712/RPCS3_PR3458_cc77c58_logs_and_screenshots.zip)
tocs_3458_cc77c58_vulkan_02

**Comment by spring568** [comment]
@Luffykun007 This might sound really basic or dumb, but did you try putting the WIP version in a different folder then moving a clean rip of the game into it? I sometimes get errors with bad config files, cache or saves created with different versions. It's less work for me doing it that way since I don't have to keep deleting a bunch of files or folders

**Comment by kd-11** [comment]
@greentop Looks like ToCS2 is leaking memory. Strict rendering mode is kinda broken right now - will investigate that issue separately, as well as the random lockups when using vulkan (and sometimes openGL)

**Comment by Psycho-A** [comment]
Still there's no shadows and it's broken lighting in Silent Hill: Downpour (uses UE3 Engine)...

**Pull request merged**